src/java/org/jivesoftware/openfire/session/LocalOutgoingServerSession.java

```diff
@@ -698,4 +700,13 @@ public String getAvailableStreamFeatures() {
         // Nothing special to add
         return null;
     }
+
+    public void addOutgoingDomainPair(String localDomain, String remoteDomain) {
+        outgoingDomainPairs.add(new DomainPair(localDomain, remoteDomain));
+    }
+
+    @Override
+    public boolean checkOutgoingDomainPair(String localDomain, String remoteDomain) {
+         return outgoingDomainPairs.contains(new DomainPair(localDomain, remoteDomain));
+    }

```

Comment:
Is there any point - or is it even safe - to let `#getAuthenticatedDomains()` exist? Isn't usage of that method inherently unsafe, in context of the issue that we're trying to fix. If so, `#addAuthenticatedDomain` can probably also go.

Reply:
Yes, you're right - I've removed both getAuthenticatedDomains and getHostnames and so on, and replaced them purely with DomainPairs throughout.